api/src/main/java/org/geysermc/geyser/api/pack/UrlPackCodec.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2019-2023 GeyserMC. http://geysermc.org
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @author GeyserMC
+ * @link https://github.com/GeyserMC/Geyser
+ */
+
+package org.geysermc.geyser.api.pack;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+/**
+ * Represents a pack codec that creates a resource
+ * pack from a URL.
+ * <p>
+ * Due to Bedrock limitations, the URL must:
+ * <ul>
+ *     <li>be a direct download link to a .zip or .mcpack resource pack</li>
+ *     <li>Use application type `application/zip` and set a correct content length</li>
+ * </ul>
+ *
+ * Additionally, the ResourcePack must be zipped in a folder enclosing the resource pack, instead of the resource pack being at the root of the zip.
+ */
+public abstract class UrlPackCodec extends PackCodec {
+
+    /**
+     * Gets the URL to the resource pack location.
+     *
+     * @return the URL of the resource pack
+     */
+    @NonNull
+    public abstract String url();
+
+    /**
+     * If the remote pack has an encryption key, it must be specified here.
+     * Otherwise, leave empty.
+     * @return the encryption key of the resource pack
+     */
+    @NonNull
+    public abstract String contentKey();
```

Comment:
I personally don't like empty strings, can we make contentKey null when there is no contentKey?

Reply:
https://github.com/GeyserMC/Geyser/blob/master/api/src/main/java/org/geysermc/geyser/api/pack/ResourcePack.java#L54-L60
Not against it; however, this is already being done the same way for other resource packs - i'd rather keep it consistent.

Reply:
I've changed it to allow setting a null content key, however, given that we use empty content keys elsewhere, I'd keep the getter as is